language-plutus-core/src/Language/PlutusCore/Clone.hs

```diff
@@ -1,50 +1,163 @@
+{-# LANGUAGE ConstraintKinds  #-}
 {-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE InstanceSigs     #-}
+{-# LANGUAGE LambdaCase       #-}
 
-module Language.PlutusCore.Clone ( cloneType
+module Language.PlutusCore.Clone (Cloneable(..), CloneWhat(..), cloneQuote, alphaRename, globalRename
                                  ) where
 
-import           Control.Monad.State.Class
-import           Control.Monad.Trans.State hiding (get, modify)
-import qualified Data.IntMap               as IM
 import           Language.PlutusCore.Name
 import           Language.PlutusCore.Quote
 import           Language.PlutusCore.Type
-import           PlutusPrelude
-
-type CloneSt = IM.IntMap (TyNameWithKind ())
-
-cloneEnvAssign :: MonadState CloneSt m
-               => Unique
-               -> TyNameWithKind ()
-               -> m ()
-cloneEnvAssign (Unique i) ty = modify (IM.insert i ty)
-
-cloneType :: MonadQuote m => Type TyNameWithKind a -> m (Type TyNameWithKind ())
-cloneType = flip evalStateT mempty . cloneTypeM
-
-replaceTyName :: (MonadQuote m, MonadState CloneSt m) => TyNameWithKind () -> m (TyNameWithKind ())
-replaceTyName (TyNameWithKind (TyName (Name l' n u))) = do
-    u' <- liftQuote freshUnique
-    let tn = void $ TyNameWithKind (TyName (Name l' n u')) -- TODO use lenses
-    cloneEnvAssign u tn
-    pure tn
-
-cloneTypeM :: (MonadQuote m, MonadState CloneSt m) => Type TyNameWithKind a -> m (Type TyNameWithKind ())
-cloneTypeM (TyFix _ tn ty) = do
-    tn' <- replaceTyName (void tn)
-    TyFix () tn' <$> cloneTypeM ty
-cloneTypeM (TyForall _ tn k ty) = do
-    tn' <- replaceTyName (void tn)
-    TyForall () tn' (void k) <$> cloneTypeM ty
-cloneTypeM (TyLam _ tn k ty) = do
-    tn' <- replaceTyName (void tn)
-    TyLam () tn' (void k) <$> cloneTypeM ty
-cloneTypeM (TyApp _ ty ty') = TyApp () <$> cloneTypeM ty <*> cloneTypeM ty'
-cloneTypeM (TyFun _ ty ty') = TyFun () <$> cloneTypeM ty <*> cloneTypeM ty'
-cloneTypeM x@(TyVar _ (TyNameWithKind (TyName (Name _ _ u)))) = do
-    cloneSt <- get
-    case IM.lookup (unUnique u) cloneSt of
-        Just n  -> pure (TyVar () n)
-        Nothing -> pure (void x)
-cloneTypeM x@TyBuiltin{}    = pure (void x)
-cloneTypeM x@TyInt{}        = pure (void x)
+
+import           Control.Applicative
+import           Control.Monad.Reader
+import           Control.Monad.State
+
+import           Data.Foldable
+import           Data.Functor.Foldable
+import qualified Data.IntMap               as IM
+import qualified Data.IntSet               as IS
+import qualified Data.List.NonEmpty        as NE
+import           Data.Maybe                (fromMaybe)
+import           Lens.Micro
+
+type Cloning m = (MonadState CloneSt m, MonadReader CloneWhat m, MonadQuote m)
+
+data CloneSt = CloneSt {
+    csMappingStack :: NE.NonEmpty (IM.IntMap Unique) -- ^ A scoped stack of unique mappings.
```

Comment:
We didn't have this explicit scoping previously, right?

Reading through the code I see that each binder introduces a new scope and binds a new variable. And variables are added to `IntMap`s only there. I.e. it looks like all thjse `IntMap`s always consists of single variables. Am I wrong here?

In any case would be nice to have some documentation on how stuff works.

Reply:
Nearly - mappings for free variables are also introduced in the scope where we first see them. So the mapping can have several elements.

However, that is in fact wrong, since free variables should always have a mapping in the global scope. I need to rework this a bit.

Reply:
I know the PR is closed, but I'm really curious whether I'm wrong about `IntMap`s always being singletons or not. What do you think?